orc8r/cloud/go/services/analytics/collector_servicer.go

```diff
@@ -18,32 +18,127 @@ import (
+	userStateMgr *calculations.UserStateMgr) *CollectorService {
+	return &CollectorService{promAPIClient: promAPIClient, calculations: calculations, analyticsConfig: analyticsConfig, userStateMgr: userStateMgr}
+}
+
+// filterResults filters the results to ensure that the user aggregate metrics get filtered out in
+// case the number of users in that context falls below minimum threshold
+func (svc *CollectorService) filterResults(results []*protos.CalculationResult) []*protos.CalculationResult {
+	userStateMgr := svc.userStateMgr
+	// nothing to filter
+	if userStateMgr == nil {
+		return results
+	}
+
+	filteredResults := []*protos.CalculationResult{}
+	analyticsConfig := svc.analyticsConfig
+	for _, result := range results {
+		// enforce the minimum user threshold constraint at network level, gateway level and deployment level
+		if metricConfig, ok := analyticsConfig.Metrics[result.GetMetricName()]; ok {
+			if metricConfig.Export {
+				if metricConfig.EnforceMinUserThreshold {
+					networkID, networkLabelPresent := result.Labels[metrics.NetworkLabelName]
+					gatewayID, gatewayLabelPresent := result.Labels[metrics.GatewayLabelName]
+
+					if networkLabelPresent {
+						userNetworkTable := userStateMgr.UsersNetworkTable[networkID]
+						if gatewayLabelPresent {
+							if numUsers, ok := userNetworkTable.UsersGatewayTable[gatewayID]; ok {
+								if numUsers > analyticsConfig.MinUserThreshold {
+									filteredResults = append(filteredResults, result)
+								} else {
+									glog.V(1).Infof("metric %s(gateway label) dropped,active users %d below user threshold %d",
+										result.GetMetricName(), numUsers, analyticsConfig.MinUserThreshold)
+								}
+							}
+						} else {
+							if userNetworkTable.TotalUsersPerNetwork > analyticsConfig.MinUserThreshold {
+								filteredResults = append(filteredResults, result)
+							} else {
+								glog.V(1).Infof("metric %s(network label) dropped,active users %d below user threshold %d",
+									result.GetMetricName(), userNetworkTable.TotalUsersPerNetwork, analyticsConfig.MinUserThreshold)
+							}
+						}
+					} else {
+						if userStateMgr.TotalUsers > analyticsConfig.MinUserThreshold {
+							filteredResults = append(filteredResults, result)
+						} else {
+							glog.V(1).Infof("metric %s(deployment) dropped, active users %d below user threshold %d", result.GetMetricName(),
+								userStateMgr.TotalUsers, analyticsConfig.MinUserThreshold)
+						}
+					}
+				} else {
+					filteredResults = append(filteredResults, result)
+				}
+			} else {
+				glog.V(1).Infof("Metric %s export configuration disabled", result.GetMetricName())
+			}
+		} else {
+			glog.Errorf("Metric Configuration not found for %s", result.GetMetricName())
+		}
+	}
+	return filteredResults
+}
+
+// registerResults registers the computed metric with Prometheus
+func (svc *CollectorService) registerResults(calcParams *calculations.CalculationParams, results []*protos.CalculationResult) {
```

Comment:
I think the registration should be done at the core orc8r service. The whole purpose of this service mesh pattern is to do as much as possible at the core service. Makes debugging easier, makes reading the code cleaner and easier, and is architectually more attractive by enforcing a clear interface between the collectors and the reporters

Reply:
As mentioned in the earlier comment. I don't feel the registration is a part of the core orc8r service. The configuration is local to each service and we use this information to determine, if the metric needs to be registered or not. The core service deals with the metric export alone. The core service's configuration doesn't have any knowledge on the metrics, it is only aware of the export configuration. It uses this and exports the metrics outside. It is possible to tie in metric registration aspect as well. I don't however see a reason to do so. It will complicate this logic than simplify it